### PR TITLE
bevy_ecs address trivial cases of `unsafe_op_in_unsafe_fn`

### PR DESCRIPTION
**Pull request by tguichaoua**
# Objective

- Part of #11590
- Fix `unsafe_op_in_unsafe_fn` for trivial cases in bevy_ecs

## Solution

Fix `unsafe_op_in_unsafe_fn` in bevy_ecs for trivial cases, i.e., add an `unsafe` block when the safety comment already exists or add a comment like "The invariants are uphold by the caller".

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Good work :) And yeah, great idea to split this out.

**Review by BD103 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Waiting for James to take a look before merging.

**Review by doonv — APPROVED** [review]

File: crates/bevy_ecs/src/bundle.rs
Comment (james7132) [inline crates/bevy_ecs/src/bundle.rs:470]: I really do not like having this in the match expression, could we lift this out to before the match expression?

File: crates/bevy_ecs/src/component.rs
Comment (james7132) [inline crates/bevy_ecs/src/component.rs:364]:
```suggestion
        // SAFETY: Contract is required to be upheld by the caller.
```

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:180]: Same with the match, having an unsafe block as the conditional in an if expression is not great for readability.

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:129]: Same with the match, having an unsafe block as the conditional in an if expression is not great for readability.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/fetch.rs:1572]:
```suggestion
                         // SAFETY: The invariants are required to be upheld by the caller.
```

This is also sort of incomplete as a safety comment, but given this PR is focused on the lint, I won't block on this.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/fetch.rs:1587]:
```suggestion
                    // SAFETY: The invariants are required to be upheld by the caller.
```

This is also sort of incomplete as a safety comment, but given this PR is focused on the lint, I won't block on this.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/fetch.rs:1089]: IMO, safety comments should not have a "see: X" indirection, as they can be very fragile if either call site or invariants change.

**Review by james7132 — CHANGES_REQUESTED** [review]
Overall, I think this is a step in the right direction given how dense some of the unsafe code is in bevy_ecs. However, I lament how it impacts end-to-end readabilty though given that some of these functions are nothing but unsafe, all the way through.

`cargo geiger` is going to explode on `bevy_ecs` from this PR though.

File: crates/bevy_ecs/src/bundle.rs
Comment (tguichaoua) [reply crates/bevy_ecs/src/bundle.rs:230]: Moving the unsafe block outside the macro expansion triggers the [unused-unsafe](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-unsafe) lint.

**Comment by tguichaoua** [comment]
> I lament how it impacts end-to-end readabilty though given that some of these functions are nothing but unsafe, all the way through.

unsafe block is a [syntactic salt](https://en.wikipedia.org/wiki/Syntactic_sugar#Syntactic_salt) to force the developer to be more careful when making changes to unsafe code. Since it is a human's responsibility to ensure the safety rules, IMHO, it's worth it to pay the cost of readability in exchange for having a well-documented code.

File: crates/bevy_ecs/src/bundle.rs
Comment (BD103) [reply crates/bevy_ecs/src/bundle.rs:230]: Perhaps it would be useful to compare compile times with and without this change? If it is faster to move the `unsafe` block out of the expansion, then I think it's ok to `#[allow(...)]` it.

**Review by james7132 — APPROVED** [review]

**Pull request merged**
